django/core/paginator.py

```diff
@@ -128,6 +131,46 @@ def _check_object_list_is_ordered(self):
 QuerySetPaginator = Paginator   # For backwards-compatibility.
 
 
+class EllipsisPaginator(Paginator):
+    """
+    Paginator with a special method, which builds page range in a compact form
+    - does not display whole range if it does not fit into given size.
+    Instead of full form, it displays beginning, middle and end parts of range
+    with ellipsis between them,
+    e.g.: 1, 2,…, 40, 41, 42, 43, 44, 45, 46, …, 49, 50.
+    Originally used in `django.contrib.admin`.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.on_each_side = kwargs.pop('on_each_side', 3)
+        self.on_ends = kwargs.pop('on_ends', 2)
+        self.max_pages_num = kwargs.pop('max_pages_num', 10)
+        super().__init__(*args, **kwargs)
+
+    def get_ellipsed_page_range(self, page_num=1):
+        if self.num_pages <= self.max_pages_num:
+            return self.page_range
+
+        # Insert "smart" pagination links, so that there are always `on_ends`
+        # links at either end of the list of pages, and there are always
+        # `on_each_side` links at either end of the "current page" link.
+        page_range = []
+        if page_num > (self.on_each_side + self.on_ends + 1):
+            page_range.extend(range(1, self.on_ends + 1))
+            page_range.append(DOT)
+            page_range.extend(range(page_num - self.on_each_side, page_num + 1))
+        else:
+            page_range.extend(range(1, page_num + 1))
+
+        if page_num < (self.num_pages - self.on_each_side - self.on_ends):
+            page_range.extend(range(page_num + 1, page_num + self.on_each_side + 1))
+            page_range.append(DOT)
+            page_range.extend(range(self.num_pages - self.on_ends + 1, self.num_pages + 1))
+        else:
+            page_range.extend(range(page_num + 1, self.num_pages + 1))
+        return page_range
```

Comment:
The original code made use of unpacking generalizations rather than excessive use of `list.append()` and `list.extend()`. Please restore this.